File: assignments/haskell/kindergarten-garden/kindergarten-garden_test.hs
import Test.HUnit (Assertion, (@=?), runTestTT, Test(..), Counts(..))
import System.Exit (ExitCode(..), exitWith)
import Garden (garden, defaultGarden, lookupPlants, Plant(..))

exitProperly :: IO Counts -> IO ()
exitProperly m = do
  counts <- m
  exitWith $ if failures counts /= 0 || errors counts /= 0 then ExitFailure 1 else ExitSuccess

testCase :: String -> Assertion -> Test
testCase label assertion = TestLabel label (TestCase assertion)

main :: IO ()
main = exitProperly $ runTestTT $ TestList
  [ TestList gardenTests ]

gardenTests :: [Test]
gardenTests =
  [ testCase "alice tests" $ do
    let alice = lookupPlants "Alice" . defaultGarden
    [Radishes, Clover, Grass, Grass] @=? alice "RC\nGG"
    [Violets, Clover, Radishes, Clover] @=? alice "VC\nRC"
  , testCase "small garden" $
    [Clover, Grass, Radishes, Clover] @=?
    lookupPlants "Bob" (defaultGarden "VVCG\nVVRC")
  , testCase "medium garden" $ do
    let g = defaultGarden "VVCCGG\nVVCCGG"
    [Clover, Clover, Clover, Clover] @=? lookupPlants "Bob" g
    [Grass, Grass, Grass, Grass] @=? lookupPlants "Charlie" g
  , testCase "full garden" $ do
    let g = defaultGarden "VRCGVVRVCGGCCGVRGCVCGCGV\nVRCCCGCRRGVCGCRVVCVGCGCV"
        l = flip lookupPlants g
    [Violets, Radishes, Violets, Radishes] @=? l "Alice"
    [Clover, Grass, Clover, Clover] @=? l "Bob"
    [Violets, Violets, Clover, Grass] @=? l "Charlie"
    [Radishes, Violets, Clover, Radishes] @=? l "David"
    [Clover, Grass, Radishes, Grass] @=? l "Eve"
    [Grass, Clover, Violets, Clover] @=? l "Fred"
    [Clover, Grass, Grass, Clover] @=? l "Ginny"
    [Violets, Radishes, Radishes, Violets] @=? l "Harriet"
    [Grass, Clover, Violets, Clover] @=? l "Ileana"
    [Violets, Clover, Violets, Grass] @=? l "Joseph"
    [Grass, Clover, Clover, Grass] @=? l "Kincaid"
    [Grass, Violets, Clover, Violets] @=? l "Larry"
  , testCase "surprise garden" $ do
    let g = garden [ "Samantha", "Patricia"
                   , "Xander", "Roger"
                   ] "VCRRGVRG\nRVGCCGCV"
        l = flip lookupPlants g
    [Violets, Clover, Radishes, Violets] @=? l "Patricia"
    [Radishes, Radishes, Grass, Clover] @=? l "Roger"
    [Grass, Violets, Clover, Grass] @=? l "Samantha"
    [Radishes, Grass, Clover, Violets] @=? l "Xander"
  ]

File: assignments/clojure/kindergarten-garden/kindergarten_garden_test.clj
(ns kindergarten.test (:require [clojure.test :refer :all]))
(load-file "garden.clj")

(deftest garden-test
  (is (= [:radishes :clover :grass :grass] (:alice (garden "RC\nGG"))))
  (is (= [:violets :clover :radishes :clover] (:alice (garden "VC\nRC")))))

(def small-garden (garden "VVCG\nVVRC"))
(deftest small-garden-test
  (is (= [:clover :grass :radishes :clover] (:bob small-garden))))

(def medium-garden (garden "VVCCGG\nVVCCGG"))
(deftest medium-garden-test
  (is (= [:clover :clover :clover :clover] (:bob medium-garden)))
  (is (= [:grass :grass :grass :grass] (:charlie medium-garden))))

(def full-garden (garden "VRCGVVRVCGGCCGVRGCVCGCGV\nVRCCCGCRRGVCGCRVVCVGCGCV"))
(deftest full-garden-test
  (is (= [:violets  :radishes :violets  :radishes] (:alice   full-garden)))
  (is (= [:clover   :grass    :clover   :clover]   (:bob     full-garden)))
  (is (= [:violets  :violets  :clover   :grass]    (:charlie full-garden)))
  (is (= [:radishes :violets  :clover   :radishes] (:david   full-garden)))
  (is (= [:clover   :grass    :radishes :grass]    (:eve     full-garden)))
  (is (= [:grass    :clover   :violets  :clover]   (:fred    full-garden)))
  (is (= [:clover   :grass    :grass    :clover]   (:ginny   full-garden)))
  (is (= [:violets  :radishes :radishes :violets]  (:harriet full-garden)))
  (is (= [:grass    :clover   :violets  :clover]   (:ileana  full-garden)))
  (is (= [:violets  :clover   :violets  :grass]    (:joseph  full-garden)))
  (is (= [:grass    :clover   :clover   :grass]    (:kincaid full-garden)))
  (is (= [:grass    :violets  :clover   :violets]  (:larry   full-garden))))

(def surprise-garden (garden "VCRRGVRG\nRVGCCGCV" ["Samantha" "Patricia" "Xander" "Roger"]))
(deftest surprise-garden-test
  (is (= [:violets  :clover   :radishes :violets] (:patricia surprise-garden)))
  (is (= [:radishes :radishes :grass    :clover]  (:roger    surprise-garden)))
  (is (= [:grass    :violets  :clover   :grass]   (:samantha surprise-garden)))
  (is (= [:radishes :grass    :clover   :violets] (:xander   surprise-garden))))

(run-tests)


